doc/manual/calling-c-and-fortran-code.rst

```diff
@@ -562,21 +564,13 @@ in the ``free`` function being called via the wrong ``libc`` library and
 cause Julia to crash. The reverse (passing an object allocated in Julia
 to be freed by an external library) is equally invalid.
 
-**Ptr{T} vs. Array{T} vs. Ref{T} vs. T**
-
-The choice of type-wrapper declaration strongly depends on who allocated the memory,
-and the declared type.
-In general, use ``T`` if the memory is intended to be allocated in
-(and managed by) Julia (with type-tag).
-Use ``Ptr{T}`` if the memory is expected to be populated by ``C`` (without type-tag).
-Use ``Ref{T}`` if you have an ``isbits`` type,
-but you want to turn it into a pointer to a struct in another struct definition.
+When to use T, Ptr{T} and Ref{T}
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-See issue #2818 for some work that needs to be done to simplify this so that Julia
-types can be used to recursively mirror c-style structs,
-without requiring as much manual management of the ``Ptr`` conversions.
-After #2818 is implemented, it will be true that an ``Vector{T}`` will be equivalent to
-an ``Ptr{Ptr{T}}``. That is currently not true, and the conversion must be explicitly.
+In C wrapper code that must handle pointers, ``Ref{T}`` should generally be
+used for the types of input arguments, whereas ``Ptr{T}`` should generally be
+used for the output type, and also within Julia immutable types designed to
+mimic the internal structure of corresponding C structs.
```

Comment:
Thanks for documenting this, that's really welcome! Could you also explain the reasons for the rules you give, i.e. the ownership differences between `Ptr` and `Ref`?